visualization/app/codeCharta/ui/artificialIntelligence/artificialIntelligenceHelper/selectors/artificialIntelligence.selector.ts

```diff
@@ -0,0 +1,103 @@
+import {
+	AREA_METRIC,
+	calculateRiskProfile,
+	EXCLUDED_FILE_EXTENSION,
+	HEIGHT_METRIC,
+	RiskProfile,
+	getPercentagesOfRiskProfile
+} from "../riskProfileHelper"
+import {
+	calculateSuspiciousMetrics,
+	findGoodAndBadMetrics,
+	setMetricValues,
+	MetricSuggestionParameters,
+	MetricValues,
+	MetricValuesByLanguage
+} from "../suspiciousMetricsHelper"
+import { hierarchy } from "d3-hierarchy"
+import { BlacklistItem, BlacklistType, CodeMapNode, NodeType } from "../../../../codeCharta.model"
+import { getMostFrequentLanguage } from "../mainProgrammingLanguageHelper"
+import { isPathBlacklisted } from "../../../../util/codeMapHelper"
+import { createSelector } from "../../../../state/angular-redux/createSelector"
+import { blacklistSelector } from "../../../../state/store/fileSettings/blacklist/blacklist.selector"
+import { experimentalFeaturesEnabledSelector } from "../../../../state/store/appSettings/enableExperimentalFeatures/experimentalFeaturesEnabled.selector"
+import { CcState } from "../../../../state/store/store"
+import { AccumulatedData, accumulatedDataSelector } from "../../../../state/selectors/accumulatedData/accumulatedData.selector"
+
+export interface ArtificialIntelligenceControllerViewModel {
+	analyzedProgrammingLanguage: string
+	suspiciousMetricSuggestionLinks: MetricSuggestionParameters[]
+	unsuspiciousMetrics: string[]
+	riskProfile: RiskProfile
+}
+
+export const calculate = (
+	experimentalFeaturesEnabled: boolean,
+	accumulatedData: Pick<AccumulatedData, "unifiedMapNode">,
+	blacklist: BlacklistItem[]
+): ArtificialIntelligenceControllerViewModel | undefined => {
+	if (!experimentalFeaturesEnabled) {
+		return
+	}
+
+	const artificialIntelligenceViewModel: ArtificialIntelligenceControllerViewModel = {
+		analyzedProgrammingLanguage: undefined,
+		suspiciousMetricSuggestionLinks: [],
+		unsuspiciousMetrics: [],
+		riskProfile: { lowRisk: 0, moderateRisk: 0, highRisk: 0, veryHighRisk: 0 }
+	}
+
+	const numberOfFilesByLanguage = new Map<string, number>()
+	const rlocRisk = { lowRisk: 0, moderateRisk: 0, highRisk: 0, veryHighRisk: 0 }
+	let totalRloc = 0
+	const metricValues: MetricValues = {}
+	const metricValuesByLanguage: MetricValuesByLanguage = {}
+
+	for (const { data } of hierarchy(accumulatedData.unifiedMapNode)) {
+		const fileExtension = getFileExtension(data.name)
+		if (data.type === NodeType.FILE && fileExtension !== undefined && !isPathBlacklisted(data.path, blacklist, BlacklistType.exclude)) {
+			const filesPerLanguage = numberOfFilesByLanguage.get(fileExtension) ?? 0
+			numberOfFilesByLanguage.set(fileExtension, filesPerLanguage + 1)
+
+			setMetricValues(data, metricValues)
+			metricValuesByLanguage[fileExtension] = metricValues
+
+			if (isFileValid(data, fileExtension)) {
+				totalRloc += data.attributes[AREA_METRIC]
+				calculateRiskProfile(data, rlocRisk, fileExtension)
+			}
+		}
+
+		if (totalRloc > 0) {
+			artificialIntelligenceViewModel.riskProfile = getPercentagesOfRiskProfile(rlocRisk)
```

Comment:
Is `getPercentagesOfRiskProfile` used elsewhere? If not, lets do not export it and call it inside calculateRiskProfile().

Reply:
We spoke about it. We leave it like that.